### ISSUE TITLE
MicroBlaze does not return from mbedtls_pk_parse_key with MBEDTLS_HAVE_ASM defined

### ISSUE THREAD

**Issue by IsaacJT** [Mbed-TLS/mbedtls#2020]
### Description
- Type: Bug 
- Priority:  Major 

---------------------------------------------------------------
## Bug

**OS**  
FreeRTOS

**mbed TLS build:**  
Version: 2.12.0
OS version: 10.0  
Configuration: 
[mbedtls_freertos.txt](https://github.com/ARMmbed/mbedtls/files/2391700/mbedtls_freertos.txt)
Compiler and options (if you used a pre-built binary, please indicate how you obtained it):  GCC 7.2.0 from Xilinx SDK 2018.1. Compiled with: `-O2 -mcpu=v10.0 -mxl-multiply-high -mno-xl-soft-div -mno-xl-soft-mul -mxl-barrel-shift -mxl-pattern-compare`
Additional environment information:  we are also using Mongoose web server. The MicroBlaze has been built with multiplier and divider hardware, branch prediction, and data and instruction caches. 

**Expected behavior**   
When parsing a buffer with `mbedtls_pk_parse_key`, the function should return.

**Actual behavior**  
The `mbedtls_pk_parse_key` never returns, and is stuck performing the multiplication in bignum.c.

The function executes and returns correctly when compiling without MBEDTLS_HAVE_ASM defined. 

**Steps to reproduce**  
Attempt to parse a key with `mbedtls_pk_parse_key` on a MicroBlaze with MBEDTLS_HAVE_ASM defined.

**Comment by IsaacJT** [comment]
The application gets stuck in the do while loop here: https://github.com/ARMmbed/mbedtls/blob/6c34268e203d23bbfbfda3f7362dac8b9b9382bc/library/bignum.c#L1407

The `mbedtls_mpi_cmp_mpi` function always returns 1.

**Comment by Patater** [comment]
Hi @IsaacJT 

Thanks for your bug report. We unfortunately don't have a way to reproduce this issue, so I've marked this as open for community contribution.

Thanks!

**Comment by Kazuyuki-Kimura** [comment]
Without knowing about this bug, I installed mbedTLS in a mail client I was making, and it stopped just like #2020, so I think I was able to reproduce it.
When I wrote a new assembler code to improve the speed and pass the test, I compared the results and found that the only difference was the order in which the 64-bit data was imported from memory.
microblaze allows you to choose the endianness when you embed it in your FPGA.
When we first released it, I think we created the code to work with big-endian cores.
We have passed all the necessary tests and confirmed that we can send emails to the SMTPS server without any problems.

**Comment by Eurecam-Benjamin** [comment]
I also runned into this issue when using mbedtls on a microblaze.
I am no microblaze guru but I further analyse the code by comparing it to the default C fallback :

- There is indeed a little/big endian problem when using "lhui" instruction -> I replaced it with "lwi" and masking / shifting value to upper and lower bits
- To me, there are also some typo with r0 register confused with 0 number.
- And last, some registers where swapped.

The attached diff file fixes include/mbedtls/nb_mul.h in branche 2.16
The lines number doesn't match but it works also with 2.28 and almost certainly with the newer 3.1 as they share the code.

[fix_bn_mul.txt](https://github.com/ARMmbed/mbedtls/files/8132763/fix_bn_mul.txt)
